### DISCUSSION TITLE
Execute some nodes when routine is manually stopped

### DISCUSSION THREAD

**Discussion by lachioma** [bonsai-rx/bonsai#1493] [Q&A]
Hello

Is it possible to run some nodes only when the Stop button is pressed, as last things to run right before stopping the program?

Thanks for any feedback!

**Comment by bruno-f-cruz** [comment]
There is currently no way of running code after the stop node is pressed.
However, see #893 and #1023 for more details.

In general, you are looking for a workflow along the lines of:
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/558f5453-bc10-411f-b2a2-df103a887cc0)

Hope it helps,
Bruno

**Comment by lachioma** [comment]
Thank you Bruno.

This indeed answer my question, but I am still not able to do what I want. Can you please try to help?

See my example routine.
I have a camera acquisition (with the camera being triggered via a TTL from an Arduino serially connected). I would like that by pressing a key, the Bonsai workflow would (1) send a string to Arduino to stop the TTL; (2) add to the video file any remaining image that might be in the buffer; (3) end the workflow. 
If in that "SelectMany" I add another `VideoWriter` node with the same filename, it will complain that the file already exists. How can I do point (2)?

Thank you for any extra help!

![image](https://github.com/bonsai-rx/bonsai/assets/29898879/b1f8f92f-0605-4705-819e-67163ac97d49)

**Comment by bruno-f-cruz** [comment]
Why do you need a videowriter inside the `SelectMany`? Sorry not sure I follow. If your TTL stops, the original videowriter will already write the frames to file.

**Comment by lachioma** [comment]
A bit more details about my Bonsai-Arduino-camera configuration.
The bonsai program sends a serial message to Arduino ('1'), which starts to send ttl pulses to trigger individual frames of the camera. When the key is pressed, Bonsai sends a message to Arduino ('0') to stop. Now, the camera outputs a TTL indicating the exposure of each frame and  I record this signal on a separate device. The number of frames that I extract from this signal is often 1 more than the frames in the video file. This makes me think that during the time I press the key to interrupt the Arduino ttl and so interrupt the camera, another frame is acquired by the camera but not saved into the video file.
I did not verify whether it is in fact that last frame to be missing from the video file, but I am sure that no frame is dropped in the middle, so probably it is the last one indeed.

**Comment by bruno-f-cruz** [comment]
It sounds like you can probably just get away with this then (like suggested in one of the posts I referenced)

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/fc2478d3-543f-45c6-85c8-a7f5c7288127)

**Reply by lachioma** [reply]
Thank you Bruno, this worked!
